pkg/controller/azvolume.go

```diff
@@ -0,0 +1,295 @@
+
+
+func (r *reconcileAzVolume) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+    _ = context.Background()
+    _ = r.Log.WithValues("azVolume", req.NamespacedName)
+
+	var azVolume v1alpha1.AzVolume
+	err := r.client.Get(ctx, request.NamespacedName, &azVolume); err != nil {
+
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		
+		// if the GET failure is triggered by other errors, log it and requeue the request
+		klog.Errorf("failed to fetch azvolume object with namespaced name %s: %v", request.NamespacedName, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	//azVolume creation
+	if azVolume.Status == nil {
+		if err := r.TriggerCreate(ctx, azVolume.Name); err != nil {
+			klog.Errorf("failed to create AzVolume (%s): %v", azVolume.Name, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+	} else if !azVolume.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.TriggerDelete(ctx , azVolume.Name); err != nil {
+			//If delete failed, requeue request
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+
+	return reconcile.Result{Requeue: true}, nil
+}
+ 
+
+func (r * reconcileAzVolume) TriggerCreate (ctx context.Context, volumeName string) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	if err := r.CreateAzVolume(ctx, azVolume); err != nil {
+		klog.Errorf("failed to create volume %s: %v", azVolume.Spec.UnderlyingVolume, err)
+		return err
+	}
+
+	// Update status of the object
+	//TODO: Should UpdateStatus happen before CreateAzVolume to "creating" or similar?? 
+	if err := r.UpdateStatus(ctx, azVolume.Name); err != nil {
+		return err
+	}
+	klog.Infof("successfully created volume (%s)and update status of AzVolume (%s)", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Name)
+	return nil
+
+}
+
+func (r * reconcileAzVolume) TriggerDelete (ctx context.Context, volumeName string) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	if err := r.DeleteAzVolume(ctx, azVolume); err != nil {
+		klog.Errorf("failed to delete volume %s: %v", azVolume.Spec.UnderlyingVolume, err)
+		return err
+	}
+
+	// Update status of the object
+	//TODO: how to handle status after deletion?  
+	if err := r.UpdateStatus(ctx, azVolume.Name); err != nil {
+		return err
+	}
+	klog.Infof("successfully deleted volume (%s)and update status of AzVolume (%s)", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Name)
+	return nil
+}
+
+
+func (r *reconcileAzVolumeAttachment) UpdateStatus(ctx context.Context, volumeName string) error {
```

Comment:
On creation, you will need to add a finalizer for this controller to the `AzVolume` object. This is what will allow the `AzVolume` controller to observe the delete request before it is deleted. On delete, remove the finalizer after the managed disk has been deleted. This can be done within `UpdateStatus` to avoid multiple updates to the object store. To support this you could pass a `bool` parameter.

See [Kubernetes / API Concepts / Resource Deletion](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-deletion) for more information on the deletion process.